[libc++] Implement P2442R1 `std::views::chunk`

shyeyian opened the pull request:
This PR implements `std::views::chunk` in header `<__ranges/chunk_view.h>`.

This is partial P2442R1 (as P2442R1 contains both `views::chunk` and `views::slide`).

Closes #171258

frederick-vs-ja commented on libcxx/include/__cxx03/module.modulemap:
```diff
@@ -1701,6 +1701,7 @@ module cxx03_std_private_ranges_all                        [system] {
 }
 module cxx03_std_private_ranges_as_rvalue_view             [system] { header "__ranges/as_rvalue_view.h" }
 module cxx03_std_private_ranges_chunk_by_view              [system] { header "__ranges/chunk_by_view.h" }
+module cxx03_std_private_ranges_chunk_view                 [system] { header "__ranges/chunk_view.h" }
```
I do believe that <code>cxx03_std_private_ranges_*meow*</code> shouldn't exist. Maybe we should at least stop adding any in this PR, and remove all existing ones in a future PR.

frederick-vs-ja commented on libcxx/test/std/ranges/range.adaptors/range.chunk/range.chunk.iter/compare.pass.cpp:
```diff
@@ -0,0 +1,44 @@
+//
+//===----------------------------------------------------------------------===//
+
+// UNSUPPORTED: c++03, c++11, c++14, c++17
```
```suggestion
// REQUIRES: std-at-least-c++23
```
Same for other new tests.

frederick-vs-ja commented on libcxx/include/__ranges/chunk_view.h:
```diff
@@ -0,0 +1,544 @@
+namespace __chunk {
+struct __fn {
+  template <viewable_range _Range>
+  [[nodiscard]] _LIBCPP_HIDE_FROM_ABI constexpr auto operator()(_Range&& __range, range_difference_t<_Range> __n) const
```
I guess it's better to use `static` `operator()`. Same for the other overload.
```suggestion
  [[nodiscard]] static _LIBCPP_HIDE_FROM_ABI constexpr auto operator()(_Range&& __range, range_difference_t<_Range> __n)
```

frederick-vs-ja reviewed (commented):
Thanks! I've merely made quick style checks as for now.

anonymouspc commented on libcxx/include/__ranges/chunk_view.h:
```diff
@@ -0,0 +1,544 @@
+
+template <view _View>
+  requires input_range<_View>
+class _LIBCPP_ABI_LLVM18_NO_UNIQUE_ADDRESS chunk_view : public view_interface<chunk_view<_View>> {
```
Ok.

frederick-vs-ja commented on libcxx/include/__ranges/chunk_view.h:
```diff
@@ -0,0 +1,544 @@
+
+  _LIBCPP_HIDE_FROM_ABI constexpr default_sentinel_t end() const noexcept { return std::default_sentinel; }
+
+  _LIBCPP_HIDE_FROM_ABI constexpr auto size()
```
Recently, we're gradually adding `[[nodiscard]]` to existing functions, so the existing style is just being superseded and I don't think it's useful to follow it. I believe it's already planned to add `[[nodiscard]]` to views in future PRs. CC @H-G-Hristov.

We should at least mark `const` versions of `begin` `[[nodiscard]]`. The situation is less clear to me for potentially mutating `begin`, but I think meaningful uses of `begin` shouldn't be discarding.

H-G-Hristov commented on libcxx/test/std/ranges/range.adaptors/range.chunk/adaptor.pass.cpp:
```diff
@@ -0,0 +1,73 @@
+  // Test `views::chunk(view, n)`
+  {
+    auto chunked = std::views::chunk(span, 2);
+    static_assert(std::same_as<decltype(chunked), std::ranges::chunk_view<std::span<int>>>);
```
The preferred pattern to test the return type is: `std::same_as<Result> decltype(auto) x = f();`, see: https://github.com/llvm/llvm-project/blob/0744da9e5af8545068fa157d4bce63c2d2436e2a/libcxx/test/std/ranges/range.adaptors/range.take.while/adaptor.pass.cpp#L58

e.g.:

```suggestion
    std::same_as<std::ranges::chunk_view<std::span<int>>>> decltype(auto) chunked = std::views::chunk(span, 2);
```

Also AFAIK it is preferred to specify the type explicitly instead of using `auto` if possible.

anonymouspc replied:
Ok.

H-G-Hristov commented on libcxx/test/std/ranges/range.adaptors/range.chunk/ctad.compile.pass.cpp:
```diff
@@ -0,0 +1,48 @@
+  static_assert(std::same_as< decltype(std::ranges::chunk_view(br, 0)),
+                              std::ranges::chunk_view<std::ranges::ref_view<borrowed_range>> >);
+  static_assert(std::same_as< decltype(std::ranges::chunk_view(std::move(br), 0)),
+                              std::ranges::chunk_view<std::ranges::owning_view<borrowed_range>> >);
```
```suggestion
  static_assert(std::same_as<decltype(std::ranges::chunk_view(std::move(r), 0)),
                              std::ranges::chunk_view<std::ranges::owning_view<range>>>);
  static_assert(std::same_as<decltype(std::ranges::chunk_view(br, 0)),
                              std::ranges::chunk_view<std::ranges::ref_view<borrowed_range>>>);
  static_assert(std::same_as<decltype(std::ranges::chunk_view(std::move(br), 0)),
                              std::ranges::chunk_view<std::ranges::owning_view<borrowed_range>>>);
```
clang-format doesn't format these because of older standards compatibility.

anonymouspc replied:
Ok!

H-G-Hristov reviewed (commented):
Thank you for working on this. This is just a quick review.

Please follow the guidelines:
https://libcxx.llvm.org/Contributing.html#pre-commit-check-list
https://libcxx.llvm.org/CodingGuidelines.html

Make sure to test `noexcept`, such as: `static_assert(noexcept(it.end()));` and for constness if applicable.

H-G-Hristov commented:
> This PR implements `std::views::chunk` in header `<__ranges/chunk_view.h>`.
> 
> This is partial P2442R1 (as P2442R1 contains both `views::chunk` and `views::slide`).
> 
> **Thank you!**

Forgot about that: The top comment is the commit message. Please put just what is relevant and what you want as a commit message in there.

anonymouspc commented on llvm/utils/gn/secondary/libcxx/include/BUILD.gn:
```diff
@@ -1352,6 +1352,7 @@ if (current_toolchain == default_toolchain) {
       "__ranges/all.h",
       "__ranges/as_rvalue_view.h",
       "__ranges/chunk_by_view.h",
+      "__ranges/chunk_view.h",
```
Ok. I'll revert it.

anonymouspc commented on libcxx/test/std/ranges/range.adaptors/range.chunk/range.chunk.iter/compare.pass.cpp:
```diff
@@ -0,0 +1,44 @@
+
+#include "test_range.h"
+
+constexpr bool test() {
```
Ok, I'll restyle all the test files into your one.

H-G-Hristov commented on libcxx/test/libcxx/ranges/range.adaptors/range.chunk/range.chunk.iterator/deref.nodiscard.verify.cpp:
```diff
@@ -0,0 +1,26 @@
+  *view.begin(); // expected-warning {{ignoring return value of function declared with 'nodiscard' attribute}}
+  *std::as_const(view).begin(); // expected-warning {{ignoring return value of function declared with 'nodiscard' attribute}}
+  // clang-format on
+}
```
In case you missed my previous comment: https://github.com/llvm/llvm-project/pull/171234#discussion_r2600958296
Unlike the general rule to test each function in a separate file, we'd put all `[[nodiscard]]` tests in a single file `nodiscard.verify.cpp`.
Currently: `libcxx/test/libcxx/diagnostics/ranges.nodiscard.verify.cpp` exists but I'd split it per range because it will become unwieldy.

`// clang-format off` and `on` is OK but I'd prefer the following to avoid having to format manually:

```suggestion
void test() {
  char range[6] = {'x', 'x', 'y', 'y', 'z', 'z'};
  auto view     = range | std::views::chunk(2);

  // expected-warning@+1 {{ignoring return value of function declared with 'nodiscard' attribute}}
  *view.begin(); 
   // expected-warning@+1 {{ignoring return value of function declared with 'nodiscard' attribute}}
  *std::as_const(view).begin();
}
```

There are also other options provided by lit but I don't use them. You can use whatever you prefer. No chances requested.

anonymouspc replied:
Ok. I've collected all the `[[nodiscard]]` tests into `libcxx/test/libcxx/ranges/range.adaptors/range.chunk/nodiscard.verify.cpp` , along with a `no_unique_address.compile.pass.cpp`.

H-G-Hristov commented on libcxx/docs/ReleaseNotes/22.rst:
```diff
@@ -49,6 +49,7 @@ Implemented Papers
 - P2835R7: Expose ``std::atomic_ref``'s object address (`Github <https://llvm.org/PR118377>`__)
 - P2944R3: Comparisons for ``reference_wrapper`` (`Github <https://llvm.org/PR105424>`__)
 - P3168R2: Give ``std::optional`` Range Support (`Github <https://llvm.org/PR105430>`__)
+- P2442R1: Add ``std::views::chunk`` (`Github <https://llvm.org/PR171234>`__)
```
We don't normally provide release notes for partially implemented papers (but the first entry does it anyway). In case somebody doesn't ask you to remove it I'd change it to:


```suggestion
- P2442R1: P2442R1: Windowing range adaptors: ``views::chunk`` and ``views::slide`` (`Github <https://llvm.org/PR171234>`__) (Implemented ``views::slide`` only)
```

I also created a sub-task for you and linked it: https://github.com/llvm/llvm-project/issues/171258 parent issue is: https://github.com/llvm/llvm-project/issues/105187


BTW. A previous stale attempts at implementing `slide` and `chunk`, which may be useful to you:

https://github.com/llvm/llvm-project/pull/85741
https://github.com/llvm/llvm-project/pull/67146

smallp-o-p commented on libcxx/include/__ranges/chunk_view.h:
```diff
@@ -0,0 +1,562 @@
+  _LIBCPP_HIDE_FROM_ABI __outer_iterator& operator=(__outer_iterator&&) = default;
+
+  [[nodiscard]] _LIBCPP_HIDE_FROM_ABI constexpr value_type operator*() const {
+    _LIBCPP_ASSERT_PEDANTIC(*this != default_sentinel, "Trying to dereference past-the-end chunk_view iterator.");
```
This should be `_LIBCPP_ASSERT_VALID_ELEMENT_ACCESS`

smallp-o-p commented on libcxx/include/__ranges/chunk_view.h:
```diff
@@ -0,0 +1,562 @@
+  [[nodiscard]] _LIBCPP_HIDE_FROM_ABI constexpr const iterator_t<_View> base() const& { return *__parent_->__current_; }
+
+  [[nodiscard]] _LIBCPP_HIDE_FROM_ABI constexpr range_reference_t<_View> operator*() const {
+    _LIBCPP_ASSERT_PEDANTIC(*this != default_sentinel, "Trying to dereference past-the-end chunk_view iterator");
```
`_LIBCPP_ASSERT_VALID_ELEMENT_ACCESS` here too

frederick-vs-ja commented:
I think errors about libunwind are completely unrelated to this PR. Let's wait for the fix and then update the branch.

anonymouspc commented:
Apologize for the noise, the previous review requests were triggered by a temporary base/branch mix-up.

The PR is now closed and reopened. Please ignore the earlier request.

Sorry.

Commit:
Implement `views::chunk`.

This commit squashes previous commits.

frederick-vs-ja commented on libcxx/include/__ranges/chunk_view.h:
```diff
@@ -0,0 +1,562 @@
+public:
+  _LIBCPP_HIDE_FROM_ABI constexpr explicit chunk_view(_View __base, range_difference_t<_View> __n)
+      : __base_(std::move(__base)), __n_(__n), __remainder_(0) {
+    _LIBCPP_ASSERT_PEDANTIC(__n > 0, "Trying to construct a chunk_view with chunk size <= 0");
```
Could you add one or more `assert.*.cpp` to `libcxx/test/libcxx/ranges/range.adaptors/range.chunk/` to test these assertions?

The test file(s) should require the following:
```C++
// REQUIRES: has-unix-headers
// REQUIRES: libcpp-hardening-mode={{extensive|debug}}
// XFAIL: libcpp-hardening-mode=debug && availability-verbose_abort-missing
```

And `TEST_LIBCPP_ASSERT_FAILURE` should be used.

shyeyian committed:
Code Review: adjust release notes

Note that we will merge this commit with the corresponding change in #172948 in the future.

shyeyian committed:
Add Tests: test if `_LIBCPP_ASSERT_XXX` works when `views::chunk` meets an undefined behaviour

shyeyian committed:
Fix CI: adjust `#include` path in `test/libcxx` files

shyeyian committed:
Fix CI: add template deduction guide for `input_span<T>`

shyeyian committed:
Code Review: add `REQUIRES: std-at-least-c++23` in all the tests related to `views::chunk`.

The previous commit, 4dc9b06c0dd515030226b9d27841352ef62549a3, which introduce manual/explicit template deduction guide in `input_span<T>`, is always implicitly-provided under -std=c++23. So this commit also reverts the changes about `types.h` in the previous commit.

anonymouspc commented on libcxx/test/libcxx/ranges/range.adaptors/range.chunk/range.chunk.iter/assert.deref.pass.cpp:
```diff
@@ -0,0 +1,52 @@
+//
+//===----------------------------------------------------------------------===//
+
+// REQUIRES: has-unix-headers
```
Ok. I've add these guardians.
Thanks!

shyeyian committed:
Fix CI: add template deduction guide for `input_span<T>` again.

Seems that even in C++23 the CI fails due to no implicit deduction guide, here we just add the explicit one and make the CI to basically run ok.

anonymouspc commented:
Hello @frederick-vs-ja,

When you have time, could you please take a look at this PR and see whether it’s ready to be merged or if any further changes are needed?

Thanks a lot — and happy new year!

H-G-Hristov commented on libcxx/test/libcxx/ranges/range.adaptors/range.chunk/nodiscard.verify.cpp:
```diff
@@ -0,0 +1,86 @@
+  // expected-warning@+1 {{ignoring return value of function declared with 'nodiscard' attribute}}
+  range | std::views::chunk(3);
+  // expected-warning@+1 {{ignoring return value of function declared with 'nodiscard' attribute}}
+  std::views::reverse | std::views::chunk(3);
```
It looks like these line are duplicated. Also I'd move these adapter tests to the bottom because they are defined last.

H-G-Hristov commented on libcxx/test/std/ranges/range.adaptors/range.chunk/base.pass.cpp:
```diff
@@ -0,0 +1,46 @@
+
+// <ranges>
+
+//   constexpr V base() const& requires copyy_constructible<V>;
```
```suggestion
//   constexpr V base() const& requires copy_constructible<V>;
```

Please verify with the synopsis.

shyeyian committed:
Code review: Update libcxx/test synopsis

Co-authored-by: Hristo Hristov <hghristov.rmm@gmail.com>

shyeyian committed:
Code review: Fix synopsis.

Co-authored-by: Hristo Hristov <hghristov.rmm@gmail.com>

shyeyian committed:
Code review: sort and unique the tests

ldionne committed:
Merge remote-tracking branch 'origin/main' into ranges_chunk

# Conflicts:
#	libcxx/include/ranges

ldionne committed:
Fix release notes

shyeyian committed:
Resolve overload conflicts on `__div_ceil` in `ranges::chunk_view` and `ranges::stride_view`.

Both `ranges::chunk_view` and `ranges::stride_view` need a helper function called `__div_ceil` that solves some simple maths. According to the style of other libcxx internal helper functions (e.g. `__priority_tag`, `__small_buffer`), it seems that we can put `__div_ceil` into a new file `__utility/div_ceil.h`. However, most files in `__utility/` refer to a corresponding symbol in `<utility>` header, so that if someone feels this is not proper, I'll be glad to revert it.

shyeyian committed:
Fix test error that `constexpr` is not supported in `-std=c++03`

shyeyian committed:
Fix test error again that `constexpr` is not supported in `-std=c++11`

shyeyian committed:
Fix test error that `std::as_const` is included from `<utility>`.

huixie90 commented on libcxx/include/__ranges/stride_view.h:
```diff
@@ -121,13 +113,13 @@ class stride_view : public view_interface<stride_view<_View>> {
     requires sized_range<_View>
   {
-    return std::__to_unsigned_like(ranges::__div_ceil(ranges::distance(__base_), __stride_));
+    return std::__to_unsigned_like(__div_ceil(ranges::distance(__base_), __stride_));
```
Please do not use ADL. `std::__div_ceil`

huixie90 commented on libcxx/include/__ranges/chunk_view.h:
```diff
@@ -0,0 +1,557 @@
+  [[nodiscard]] _LIBCPP_HIDE_FROM_ABI constexpr auto size()
+    requires sized_range<_View>
+  {
+    return std::__to_unsigned_like(__div_ceil(ranges::distance(__base_), __n_));
```
avoid ADL. `std::__div_ceil`

huixie90 commented on libcxx/include/__ranges/chunk_view.h:
```diff
@@ -0,0 +1,557 @@
+  [[nodiscard]] _LIBCPP_HIDE_FROM_ABI constexpr auto size() const
+    requires sized_range<const _View>
+  {
+    return std::__to_unsigned_like(__div_ceil(ranges::distance(__base_), __n_));
```
same as above

huixie90 commented on libcxx/test/std/ranges/range.adaptors/range.chunk/range.chunk.iter/compare.pass.cpp:
```diff
@@ -0,0 +1,108 @@
+  std::ranges::chunk_view<std::ranges::ref_view<std::vector<int>>> chunked = vector | std::views::chunk(3);
+  std::ranges::chunk_view<input_span<int>> input_chunked = input_span<int>(vector) | std::views::chunk(3);
+
+  // Test `friend constexpr bool opreator==(const outer_iterator& x, default_sentinel_t)`
```
I have not look at the tests in details. But I think the coverage is bit insufficient. Could you please add tests for the cases

- All the constraints need to be tested (with static_assert and concepts)
- negative cases. (for example, for this test `==` we also need to test when it does not equal to default sentinel)
- empty ranges
- a range with only 1 element
- a range with more than 1 element
- fully divisible ranges e.g. chunk of 3 with 12 elements
- not fully divisible ranges chunk of 3 with 11 elements
- chunk of 1
- chunk of N > 1
- chunk of N > size of the range
- different types of underlying iterators , we have lots of iterator types in test_iterators.h and test_ranges.h, and usually we template the tests with different types of the iterators/sentinel types

These comments applies to all of the tests

huixie90 commented on libcxx/test/std/ranges/range.adaptors/range.chunk/range.chunk.iter/compare.pass.cpp:
```diff
@@ -0,0 +1,108 @@
+  std::vector<int> vector                                                  = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12};
+  std::ranges::chunk_view<std::ranges::ref_view<std::vector<int>>> chunked = vector | std::views::chunk(3);
+  std::ranges::chunk_view<input_span<int>> input_chunked = input_span<int>(vector) | std::views::chunk(3);
+
```
for this particular test, I think we need to test the branch where 

__parent_->__remainder_ == 0;

huixie90 commented on libcxx/include/__ranges/chunk_view.h:
```diff
@@ -0,0 +1,557 @@
+
+public:
+  class value_type;
+  using iterator_concept = input_iterator_tag;
```
do we have tests for these typedefs and types?

huixie90 commented on libcxx/test/std/ranges/range.adaptors/range.chunk/range.chunk.iter/compare.pass.cpp:
```diff
@@ -0,0 +1,108 @@
+
+// <ranges>
+
+//   V models only input_range
```
have you considered split the tests between the input-only case and the rest cases as the spec has two different instantiations. would that make tests more readable?

huixie90 commented on libcxx/include/__ranges/chunk_view.h:
```diff
@@ -0,0 +1,557 @@
+  using iterator_concept = input_iterator_tag;
+  using difference_type  = range_difference_t<_View>;
+
+  _LIBCPP_HIDE_FROM_ABI __outer_iterator(__outer_iterator&&) = default;
```
do we miss a test for those functions?

huixie90 commented on libcxx/include/__ranges/chunk_view.h:
```diff
@@ -0,0 +1,557 @@
+    const auto __dist = ranges::end(__i.__parent_->__base_) - *__i.__parent_->__current_;
+    if (__dist < __i.__parent_->__remainder_)
+      return __dist == 0 ? 0 : 1;
+    return __div_ceil(__dist - __i.__parent_->__remainder_, __i.__parent_->__n_) + 1;
```
avoid ADL

huixie90 commented on libcxx/include/__ranges/chunk_view.h:
```diff
@@ -0,0 +1,557 @@
+
+template <view _View>
+  requires input_range<_View>
+class chunk_view<_View>::__outer_iterator::value_type : public view_interface<value_type> {
```
I think the test for this class is missing

huixie90 commented on libcxx/test/std/ranges/range.adaptors/range.chunk/range.chunk.iter/compare.pass.cpp:
```diff
@@ -0,0 +1,108 @@
+
+//   V models only input_range
+//     friend constexpr bool operator==(const outer_iterator& x, default_sentinel_t);
+//     friend constexpr bool operator==(const inner_iterator& x, default_sentinel_t);
```
I think we should split the inner iterator and outer iterator tests

huixie90 commented on libcxx/include/__ranges/chunk_view.h:
```diff
@@ -0,0 +1,557 @@
+      : __parent_(std::addressof(__parent)) {}
+
+public:
+  using iterator_concept = input_iterator_tag;
```
do we have tests for these typedefs ?

huixie90 commented on libcxx/include/__ranges/chunk_view.h:
```diff
@@ -0,0 +1,557 @@
+  using difference_type  = range_difference_t<_View>;
+  using value_type       = range_value_t<_View>;
+
+  _LIBCPP_HIDE_FROM_ABI __inner_iterator(__inner_iterator&&) = default;
```
do we have tests for the special member functions?

huixie90 commented on libcxx/include/__ranges/chunk_view.h:
```diff
@@ -0,0 +1,557 @@
+
+  _LIBCPP_HIDE_FROM_ABI __inner_iterator& operator=(__inner_iterator&&) = default;
+
+  [[nodiscard]] _LIBCPP_HIDE_FROM_ABI constexpr const iterator_t<_View> base() const& { return *__parent_->__current_; }
```
missing tests?

huixie90 commented on libcxx/include/__ranges/chunk_view.h:
```diff
@@ -0,0 +1,557 @@
+    return -(__s - __i);
+  }
+
+  [[nodiscard]] _LIBCPP_HIDE_FROM_ABI friend constexpr auto
```
missing tests?

huixie90 replied:
the spec's return type is `range_rvalue_reference_t<V>` , which is not necessarily the same as `auto`, which always decays.

huixie90 commented on libcxx/include/__ranges/chunk_view.h:
```diff
@@ -0,0 +1,557 @@
+    return ranges::iter_move(*__i.__parent_->__current_);
+  }
+
+  _LIBCPP_HIDE_FROM_ABI friend constexpr void
```
same as above

huixie90 commented on libcxx/include/__ranges/chunk_view.h:
```diff
@@ -0,0 +1,557 @@
+  requires forward_range<_View>
+class chunk_view<_View> : public view_interface<chunk_view<_View>> {
+  _LIBCPP_NO_UNIQUE_ADDRESS _View __base_;
+  _LIBCPP_NO_UNIQUE_ADDRESS range_difference_t<_View> __n_;
```
question: would this _LIBCPP_NO_UNIQUE_ADDRESS make any difference to `__n_`

huixie90 commented on libcxx/include/__ranges/chunk_view.h:
```diff
@@ -0,0 +1,557 @@
+  class __iterator;
+
+public:
+  _LIBCPP_HIDE_FROM_ABI constexpr explicit chunk_view(_View __base, range_difference_t<_View> __n)
```
Do we have test for this public ctor

huixie90 commented on libcxx/include/__ranges/chunk_view.h:
```diff
@@ -0,0 +1,557 @@
+  }
+
+  [[nodiscard]] _LIBCPP_HIDE_FROM_ABI constexpr _View base() const&
+    requires std::copy_constructible<_View>
```
nit: in other places in this patch it was `copy_constructible` instead of `std::copy_constructible`. for consistency, might just remove this std::

huixie90 commented on libcxx/include/__ranges/chunk_view.h:
```diff
@@ -0,0 +1,557 @@
+  [[nodiscard]] _LIBCPP_HIDE_FROM_ABI constexpr auto size()
+    requires sized_range<_View>
+  {
+    return std::__to_unsigned_like(__div_ceil(ranges::distance(__base_), __n_));
```
```suggestion
    return std::__to_unsigned_like(std::__div_ceil(ranges::distance(__base_), __n_));
```

huixie90 commented on libcxx/include/__ranges/chunk_view.h:
```diff
@@ -0,0 +1,557 @@
+  [[nodiscard]] _LIBCPP_HIDE_FROM_ABI constexpr auto size() const
+    requires sized_range<const _View>
+  {
+    return std::__to_unsigned_like(__div_ceil(ranges::distance(__base_), __n_));
```
```suggestion
    return std::__to_unsigned_like(std::__div_ceil(ranges::distance(__base_), __n_));
```

huixie90 commented on libcxx/include/__ranges/chunk_view.h:
```diff
@@ -0,0 +1,557 @@
+  }
+
+public:
+  using iterator_category = input_iterator_tag;
```
please add tests for these typdef

huixie90 commented on libcxx/include/__ranges/chunk_view.h:
```diff
@@ -0,0 +1,557 @@
+    return *this;
+  }
+
+  _LIBCPP_HIDE_FROM_ABI constexpr __iterator operator--(int) {
```
missing constraints. That probably means that the test that tests the constraints are also missing

huixie90 commented on libcxx/include/__ranges/chunk_view.h:
```diff
@@ -0,0 +1,557 @@
+      ranges::advance(__current_, __n_ * (__x - 1));
+      __missing_ = ranges::advance(__current_, __n_, __end_);
+    } else if (__x < 0) {
+      ranges::advance(__current_, __n_ * __x + __missing_);
```
i think the test for this branch is missing

huixie90 reviewed (changes requested):
Thanks for the patch. I had a quick review on the source code and it mostly looks good. I put some comments on the source code. I have not reviewed the tests yet but I have put some high level comments on the test. Thanks a lot

shyeyian committed:
Avoid ADL in `std::__div_ceil`.

shyeyian committed:
[Important] Add missing concepts requirements.

Zingam commented on libcxx/include/__ranges/chunk_view.h:
```diff
@@ -0,0 +1,558 @@
+
+  _LIBCPP_HIDE_FROM_ABI constexpr void operator++(int) { ++*this; }
+
+  [[nodiscard]] _LIBCPP_HIDE_FROM_ABI friend constexpr bool
```
```suggestion
  _LIBCPP_HIDE_FROM_ABI friend constexpr bool
```

https://libcxx.llvm.org/CodingGuidelines.html#apply-nodiscard-where-relevant

See: 
> `[[nodiscard]]` should not be applied to functions if Clang already diagnoses unused results, for example: - conversion functions - equality operators - relational operators

Zingam commented on libcxx/include/__ranges/chunk_view.h:
```diff
@@ -0,0 +1,558 @@
+      _Parent* __parent, iterator_t<_Base> __current, range_difference_t<_Base> __missing = 0)
+      : __current_(__current), __end_(ranges::end(__parent->__base_)), __n_(__parent->__n_), __missing_(__missing) {}
+
+  [[nodiscard]] static consteval auto __get_iterator_concept() {
```
```suggestion
 static consteval auto __get_iterator_concept() {
```

We not apply `[[nodiscard]]` to internal functions usually.

Zingam commented on libcxx/include/__ranges/chunk_view.h:
```diff
@@ -0,0 +1,558 @@
+    return *this += -__x;
+  }
+
+  [[nodiscard]] _LIBCPP_HIDE_FROM_ABI friend constexpr bool operator==(const __iterator& __x, const __iterator& __y) {
```
```suggestion
  _LIBCPP_HIDE_FROM_ABI friend constexpr bool operator==(const __iterator& __x, const __iterator& __y) {
```
Reminder: https://github.com/llvm/llvm-project/pull/171234/changes#r3479366689

Same as above and please fixe the operators bellow.

Zingam commented on libcxx/test/libcxx/ranges/range.adaptors/range.chunk/nodiscard.verify.cpp:
```diff
@@ -0,0 +1,73 @@
+  std::as_const(view).begin();
+
+  // expected-warning@+1 {{ignoring return value of function declared with 'nodiscard' attribute}}
+  std::ranges::iter_move(view.begin());
+  // expected-warning@+1 {{ignoring return value of function declared with 'nodiscard' attribute}}
+  std::ranges::iter_move(std::as_const(view).begin());
```
This tests `std::ranges::iter_move` and not the member `iter_move`. 
Maybe not te best example but you want something like that:
https://github.com/llvm/llvm-project/blob/8d60471f89fc7b57c08d64d7524dfbff48078567/libcxx/test/libcxx/ranges/range.adaptors/range.join.with/nodiscard.verify.cpp#L44
Also there is no need to write to test cases if one is enough.

e.g. 

```suggestion

  auto cIt = std::as_const(view).begin();

....

  // expected-warning@+1 {{ignoring return value of function declared with 'nodiscard' attribute}}
  iter_move(cIt);
```

Zingam commented on libcxx/test/libcxx/ranges/range.adaptors/range.chunk/nodiscard.verify.cpp:
```diff
@@ -0,0 +1,73 @@
+  // expected-warning@+1 {{ignoring return value of function declared with 'nodiscard' attribute}}
+  std::views::chunk(range, 3);
+  // expected-warning@+1 {{ignoring return value of function declared with 'nodiscard' attribute}}
+  range | std::views::chunk(3);
+  // expected-warning@+1 {{ignoring return value of function declared with 'nodiscard' attribute}}
+  std::views::reverse | std::views::chunk(3);
```
```suggestion
```

These test `operator|`. It is tested elsewhere. Please remove.

Zingam commented on libcxx/test/libcxx/ranges/range.adaptors/range.chunk/nodiscard.verify.cpp:
```diff
@@ -0,0 +1,73 @@
+  // expected-warning@+1 {{ignoring return value of function declared with 'nodiscard' attribute}}
+  (view.begin() == std::as_const(view).end());
+  // expected-warning@+1 {{ignoring return value of function declared with 'nodiscard' attribute}}
+  (std::as_const(view).begin() == std::as_const(view).end());
```
```suggestion
```

Please remove. In fact you have added many more `[[nodiscard]]`s that are untested but they need to be reverted as per: https://libcxx.llvm.org/CodingGuidelines.html#apply-nodiscard-where-relevant

Zingam reviewed (changes requested):
I did a quick review of the `[[nodiscard]]` application and the related tests cases. 
These are tracked here: https://github.com/llvm/llvm-project/issues/172124 but since this is a new PR we don't need to track it there but you can use it as a reference.

All comments bellow are also relevant to: https://github.com/llvm/llvm-project/pull/172948
Please update both PRs so I don't have to bother you with the same comments twice.

Also note that most `[[nodiscard]]` tests are implemented in a style to ease the review, which means all test cases were added in the test in the same order as they appear in the header and also if split in multiple files ordered by files.

FYI: Please also make sure to "complete" your PR and get the CI green before you ask for a review as some reviewers might not review it if the PR is draft or the CI has errors.

shyeyian committed:
Cover tests on `chunk_range`. (I'll continue to cover tests on `__iterator`, `__inner_iterator` and `__outer_iterator` at the weekend)

shyeyian committed:
Fix chunk_view ctor test for forward ranges.

shyeyian committed:
Use consistent copy_constructible spelling in chunk_view

shyeyian committed:
Remove unnecessary no_unique_address from chunk_view counts

shyeyian committed:
Fix chunk_view inner iter_move return type (from T into T&&)

shyeyian committed:
Add chunk_view constraint tests

shyeyian committed:
Add chunk_view iterator type tests

shyeyian committed:
Add chunk_view inner iterator special member tests

shyeyian committed:
Add chunk_view inner iterator base() test

shyeyian committed:
Fix input_span operator* to return a reference

shyeyian committed:
Add chunk_view inner iterator iter_swap test

shyeyian committed:
Add chunk_view outer iterator value_type size test

shyeyian committed:
Add chunk_view outer iterator special member tests

shyeyian committed:
Add chunk_view forward iterator base() test

shyeyian committed:
Add chunk_view forward iterator converting constructor test

shyeyian committed:
Add negative comparison cases for chunk_view sentinel equality

shyeyian committed:
Add chunk_view forward iterator comparison constraints and negative cases

shyeyian committed:
Add chunk_view iterator arithmetic test for partial final chunk

shyeyian committed:
Remove [[nodiscard]] from chunk_view comparison operators

shyeyian committed:
Simplify chunk_view nodiscard.verify.cpp

shyeyian committed:
Split chunk_view compare.pass.cpp by iterator kind

shyeyian committed:
Split chunk_view deref.pass.cpp by iterator kind

shyeyian committed:
Split chunk_view increment.pass.cpp by iterator kind

shyeyian committed:
Split chunk_view arithmetic.pass.cpp by iterator kind

shyeyian committed:
Adjust test 'title' style

shyeyian committed:
Add [[nodiscard]] tests for full coverage

shyeyian committed:
Merge remote-tracking branch 'upstream/main' into ranges_chunk

shyeyian committed:
Fix some typo

shyeyian committed:
Cover partial-chunk branches in outer/inner iterator arithmetic and size()

shyeyian committed:
Add exact return-type checks for chunk iterator operator*/operator[]

shyeyian committed:
Remove useless tests

shyeyian committed:
Remove not-needed `#include`s

shyeyian committed:
Adjust code styles

shyeyian committed:
clang-format

ldionne commented on libcxx/docs/ReleaseNotes/23.rst:
```diff
@@ -55,6 +55,7 @@ Implemented Papers
 - P3369R0: constexpr for ``uninitialized_default_construct`` (`Github <https://llvm.org/PR118380>`__)
 - P3508R0: Wording for "constexpr for specialized memory algorithms" (`Github <https://llvm.org/PR118379>`__)
 - P4206R0: Revert string support in ``std::constant_wrapper`` (`Github <https://llvm.org/PR203338>`__)
+- P2442R1: Windowing range adaptors: ``views::chunk`` and ``views::slide`` (`Github <https://llvm.org/PR171234>`__) (Implemented ``views::chunk`` only)
```
This should move to 24.rst now.

ldionne committed:
Merge branch 'main' into ranges_chunk

ldionne committed:
Move release notes

Zingam commented on libcxx/test/std/ranges/range.adaptors/range.chunk/range.chunk.inner.iter/iter_move.pass.cpp:
```diff
@@ -0,0 +1,43 @@
+  static_assert(test());
+
+  return 0;
+}
```
It looks like you are missing an empty line at the end of the file here.

shyeyian committed:
remove `input_span` from arithmetic test

shyeyian committed:
adjust comments

shyeyian committed:
remove `types.h` from base.pass

shyeyian committed:
remove `types.h` from compare.pass

shyeyian committed:
update tests for ctor and make it clear

shyeyian committed:
update deref.pass

shyeyian committed:
update increment.pass

shyeyian committed:
update iter_move.pass

shyeyian committed:
update iter_swap.pass

shyeyian committed:
update types.compile.pass

shyeyian committed:
update arithmetic.pass

shyeyian committed:
update base.pass

shyeyian committed:
update compare.pass

shyeyian committed:
update ctor.compile.pass

shyeyian committed:
update decrement.pass

shyeyian committed:
update deref.pass

shyeyian committed:
update increment.pass

shyeyian committed:
update subscript.pass

shyeyian committed:
update types.compile.pass

shyeyian committed:
update arithmetic.pass

shyeyian committed:
update ctor.compile.pass

shyeyian committed:
fix ctor.compile.pass

shyeyian committed:
update deref.pass

shyeyian committed:
update increment.pass

shyeyian committed:
update size.pass

shyeyian committed:
update types.compile.pass and unify style

shyeyian committed:
adjust comments

shyeyian committed:
clang-format

shyeyian committed:
add operator+=(-n) branch testf for iterator

shyeyian committed:
update arithmetic tests to cover both general and empty branch
